### DISCUSSION TITLE
Allow to define multiple slots with the same name to render a single fill in multiple places

### DISCUSSION THREAD

**Discussion by JuroOravec** [django-components/django-components#419] [Ideas]
As part of https://github.com/EmilStenstrom/django-components/pull/416, I came across this snippet:

https://github.com/EmilStenstrom/django-components/blob/master/src/django_components/component.py#L332

```py
if slot_name in slot_name2fill_content:
    raise TemplateSyntaxError(
        f"Slot name '{slot_name}' re-used within the same template. "
        f"Slot names must be unique."
        f"To fix, check template '{template.name}' "
        f"of component '{self.registered_name}'."
    )
```

Basically it prevents multiple slots with the same name to be defined in a template, e.g.:

```django
1
{% slot "myslot" %}{% endslot %}
2
{% slot "myslot" %}{% endslot %}
3
```

At my work, we've hit this error when we wanted to render a single slot fill in different places based on some conditions, e.g.:

```django
{% if myvar %}
  ...
  {% slot "myslot" %}{% endslot %}
  ...
{% else %}
  ...
  {% slot "myslot" %}{% endslot %}
  ...
{% endif %}
```

Intuition tells me that this should not be an error. Basically, if I define a slot in multiple places, the fill should simply populate all the slots, so the first example would render:

```django
1
hello from myslot
2
hello from myslot
3
```

Was there a specific reason why a slot multiple times leads to errors?

I haven't looked into it, but my guess is that it throws an error because it's simply not implemented. Because when I consider the example above, then the way slots are managed behind the scenes probably needs to be refactored if the slots are to allow supporting re-use.

Simply consider this example, where we use a single slot twice, but in each case it has a different default content:

```django
{% slot "myslot" %}
  default slot 1
{% endslot %}

{% slot "myslot" %}
  default slot 2
{% endslot %}
```

**Comment by EmilStenstrom** [comment]
I'm a little hazy on the details here, but I think this has to do with nested components needing to have unique slot names amongst themselves?

```django
{% component "card" %}
  {% fill "header" %}  <!-- Error? -->
    {% component "profile" %}
      {% fill "header" %}  <!-- Error? -->
        Title
      {% endfill %}
    {% endcomponent %}  
  {% endfill %}
{% endcomponent %}
```

I'm a little worried that components with dynamic slots (slots that might not exist depending on some if statements) like the conditional one you describe above will be hard to understand and reason about. 

Could your example be rewritten as:

```django
{% slot "myslot" %}
  {% if myvar %}
    ...
  {% else %}
   ...
  {% endif %}
{% endslot %}
```

**Reply by JuroOravec** [reply]
Heya, I'm just getting to this again.

1. I think this is a different issue, because for what I have in mind, you could have only 1 slot fill. To give a better idea, imagine following example - I have a page and I want to include website logo both in the header and in the footer. So I define the `{% slot %}` with the same name twice.

	```django
	{% load component_tags %}
	<body>
	    <header>
			{% slot "logo" %}{% endslot %}
			Bla bla
		</header>
	    <main>
			{% slot "main" %} Default main {% endslot %}
		</main>
	    <footer>
			{% slot "logo" %}{% endslot %}
			Footer footer
		</footer>
	</body>
	```

	And then render it by defining the slot fill once:
	```django
	{% component "page" %}
		{% fill "logo" %}
			<img src="....">
		{% endfill %}
		{% fill "main" %}
			<article>
				Bla bla....
			</article>
		{% endfill %}
	{% endcomponent %}
	```

2. I played around, and enabling slot reuse is as simple as removing the error here https://github.com/JuroOravec/django-components/blob/d9e7e2890d97c850555cd11286081ee0519f5bf2/src/django_components/slots.py#L421.

3. As for an error with nested slot fills, this should be now resolved thanks to https://github.com/EmilStenstrom/django-components/pull/437. Previously slot fills were cached per template, which could cause the error you've shown. Now the slot fills are scoped per their respective component nodes (the `{% component %}` block they are in).

	So while this should now be fine:
	```django
	{% component "card" %}
	  {% fill "header" %}
	    {% component "profile" %}
	      {% fill "header" %}
	        Title
	      {% endfill %}
	    {% endcomponent %}  
	  {% endfill %}
	{% endcomponent %}
	```

	This should still be an error/undefined behavior:
	```django
	{% component "card" %}
	  {% fill "header" %}  <!-- Error? -->
		123
	  {% endfill %}
	  {% fill "header" %}  <!-- Error? -->
		456
	  {% endfill %}
	{% endcomponent %}
	```

---

Update: Ok, I did write a test with nested slot fills with the same name, and found there's a bug there. Altho slightly different than what we've shown here - Slot fills of inner component override the OTHER slots of the outer component.

Using the template from `slotted_template.html`:

```django
{% load component_tags %}
{% component "test" name='Igor' %}
    {% fill "header" %}  <!-- This fill is rendered correctly -->
        {% component "test" name='Joe2' %}
            {% fill "header" %} <!-- This fill is rendered correctly -->
                Name2: {{ name }}
            {% endfill %}
            {% fill "main" %}  <!-- This fill is applied also to the outer component -->
                Day2: {{ day }}
            {% endfill %}
            {% fill "footer" %}  <!-- This fill is applied also to the outer component -->
                XYZ
            {% endfill %}
        {% endcomponent %}
    {% endfill %}

	<!-- ERROR HERE: explicit "footer" slot fill is ignored because of the inner slot fill -->
    {% fill "footer" %}
        WWW
    {% endfill %}

    <!-- ERROR HERE: the default "main" slot fill is ignored because of the inner slot fill -->
{% endcomponent %}
```

I wrote a ticket for this issue here https://github.com/EmilStenstrom/django-components/issues/453

Note that this issue is NOT related to nor blocker for the slot reuse (the original topic of this thread).

**Comment by JuroOravec** [comment]
Implemented in https://github.com/EmilStenstrom/django-components/issues/453
